### DISCUSSION TITLE
Seeming race condition with @layer directive

### DISCUSSION THREAD

**Discussion by baffalop** [tailwindlabs/tailwindcss#9938] [Help] [answered]
I have a project using Tailwind 3.2, Vite 3.1 and Vue 3.2. I have defined a handful of classes inside `@layer components`, and I'm using these both in a global stylesheet, assigning directly to elements in Vue files, and via `@apply`. All of that works on the whole, but every once in a while when I change some styles (particularly those involving `@apply` of my custom classes) I get an error that one of my custom classes is not defined.

```
[postcss] /opt/build/repo/src/styles.css:33:3: The `filter-accent` class does not exist. If `filter-accent` is a custom class, make sure it is defined within a `@layer` directive.
```

If the error happens during watch, I have to rerun watch and eventually it succeeds. Sometimes it happens during build as well, and again rerunning it usually works.

My CSS is structured as such: my `@layer` is colocated with the `@tailwind` directives in a `tailwind.css` file. I have a global stylesheet `styles.css` referencing these via `@apply`. Both of these files are imported in my `main.ts`, which is the file that initialises Vue.

`tailwind.css`:  
```css
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
  .filter-accent {
    @apply filter -hue-rotate-30 brightness-70;
  }

  /* etc */
}
```

`main.ts`:  
```ts
import { createApp } from 'vue'

import '@/tailwind.css'
import '@/styles.css'

import App from '@/App.vue'

// more setup

const app = createApp(App)
app.mount('#app')
```

I've also tried putting my `@layer` directives elsewhere, eg. in the `styles.css` or in `App.vue`. In general I found that moving that code would result in an error that I needed to declare `@tailwind components` before I could use `@layer components`...

Have I made a mistake in the way I've structured these files, such that the load order is not guaranteed?

**Comment by wongjn** [comment]
You could try having the `tailwind.css` as the main CSS "entry" file, and have the CSS from `styles.css` in the `tailwind.css` file or have `styles.css` `@import`-ed within the the `tailwind.css`.

**Reply by baffalop** [reply]
This was actually a good suggestion. I did it the other way round, `@import`ing `tailwind.css` inside `styles.css`, but this appears to have sufficiently alleviated the flakiness of the build.

**Reply by baffalop** [reply]
Update: maybe not so much... now postcss is failing on one of my Vue components instead of `styles.css`

**Comment by adamwathan** [comment, marked as answer]
Hey! You're right about this being a race condition, I've tried to explain what's actually happening in the docs here:

https://tailwindcss.com/docs/functions-and-directives#using-apply-with-per-component-css

It basically boils down to the fact that bundlers treat every single `<style>` block in your Vue components as completely separate CSS entry points, and they are all processed completely independently. The bundler is actually running Tailwind separately for every single one, so if you have 50 components with `<style>` blocks, Tailwind is being run 50 times.

We don't have any control over how this happens or what order it happens in, so there's a race condition where the bundler might compile one of your `<style>` blocks that has an `@apply` in it with a custom class _before_ it compiles the CSS file where you've actually defined that class, and that's what causes the error you're seeing. We share a bunch of state for all of these separate builds for performance reasons (since they use the same config) so that's why it sometimes works and sometimes doesn't.

Best solution like mentioned in those docs I linked is to define any custom CSS as plugins in your config if you want to be able to reference that stuff across multiple CSS entry points.

**Reply by baffalop** [reply]
Ahh, thank you! I had a quick scan over that page to see if it had instructions about importing css files in a certain order, but somehow didn't dive far enough to see that part.

I'm just curious about this final note:

>Honestly though the best solution is to just not do weird stuff like this at all. Use Tailwind’s utilities directly in your markup the way they are intended to be used, and don’t abuse the `@apply` feature to do things like this and you will have a much better experience.

Does what I'm doing count as "weird stuff"/abuse of the `@apply` feature? I found it quite handy to define a (quite small) handful of classes which bundle together styles I want to reuse throughout my app, and to be able to use them as I would any other Tailwind utility class. If that's not the use case for `@layer`/`@apply`, what is?

**Reply by baffalop** [reply]
On second thoughts @adamwathan I'm still a little confused why this race condition affects even my global `styles.css`, which is imported in a typescript file and doesn't (I would expect) participate in Vue's SFC compilation?

That is, my `@layer` declarations are in `tailwind.css`, and they are used in `@apply` in `styles.css`. Both are imported in `index.ts`. This is all outside of any Vue component definitions. While I have occasionally encountered a postcss error relating to a Vue SFC, I much more frequently encounter an error relating to `styles.css`.

**Reply by adamwathan** [reply]
> That is, my @layer declarations are in tailwind.css, and they are used in @apply in styles.css. Both are imported in index.ts.

Can you share a project that reproduces this? That should be fine for sure.

**Reply by adamwathan** [reply]
Ah never mind I misread — this is the same problem. When you import the files in JS, they become separate entry points and are processed separately. If you instead just have a single CSS file that imports your other CSS files, things will work the way you expect.

TL;DR, if you have multiple CSS files, combine them with CSS `@import`, not by importing them separately into your JS files.
